### ISSUE TITLE
Errors on run : Unhandled Java exception

### ISSUE THREAD

**Issue by tomstuder** [ruby-processing/JRubyArt#46]
I feel this is an installation or compatability issue, please confirm.

**Error Message on run:**
```
tstuder:contributed tstuder$ k9 --run fire.rb 
Unhandled Java exception: java.lang.NoClassDefFoundError: com/apple/eawt/QuitHandler
java.lang.NoClassDefFoundError: com/apple/eawt/QuitHandler
                getDeclaredMethods0 at java/lang/Class.java:-2
          privateGetDeclaredMethods at java/lang/Class.java:3119
                getMethodsRecursive at java/lang/Class.java:3260
                         getMethod0 at java/lang/Class.java:3246
                          getMethod at java/lang/Class.java:2065
                          runSketch at processing/core/PApplet.java:10751
                          runSketch at processing/core/PApplet.java:10989
                          runSketch at processing/core/PApplet.java:10995
                            invoke0 at jdk/internal/reflect/NativeMethodAccessorImpl.java:-2
                             invoke at jdk/internal/reflect/NativeMethodAccessorImpl.java:62
                             invoke at jdk/internal/reflect/DelegatingMethodAccessorImpl.java:43
                             invoke at java/lang/reflect/Method.java:564
  invokeDirectWithExceptionHandling at org/jruby/javasupport/JavaMethod.java:440
                 tryProxyInvocation at org/jruby/javasupport/JavaMethod.java:625
                       invokeDirect at org/jruby/javasupport/JavaMethod.java:301
                         initialize at /Users/tstuder/.rvm/gems/ruby-2.5.1/gems/jruby_art-1.5.2/lib/jruby_art/app.rb:116
                        newInstance at org/jruby/RubyClass.java:997
                               call at org/jruby/RubyClass$INVOKER$i$newInstance.gen:-1
                load_and_run_sketch at /Users/tstuder/.rvm/gems/ruby-2.5.1/gems/jruby_art-1.5.2/lib/jruby_art/runners/base.rb:46
   invokeOther2:load_and_run_sketch at Users/tstuder/$_dot_rvm/gems/ruby_minus_2_dot_5_dot_1/gems/jruby_art_minus_1_dot_5_dot_2/lib/jruby_art/runners//Users/tstuder/.rvm/gems/ruby-2.5.1/gems/jruby_art-1.5.2/lib/jruby_art/runners/run.rb:4
                             <main> at Users/tstuder/$_dot_rvm/gems/ruby_minus_2_dot_5_dot_1/gems/jruby_art_minus_1_dot_5_dot_2/lib/jruby_art/runners//Users/tstuder/.rvm/gems/ruby-2.5.1/gems/jruby_art-1.5.2/lib/jruby_art/runners/run.rb:4
                invokeWithArguments at java/lang/invoke/MethodHandle.java:707
                          runScript at org/jruby/Ruby.java:852
                        runNormally at org/jruby/Ruby.java:771
                        runNormally at org/jruby/Ruby.java:789
                        runFromMain at org/jruby/Ruby.java:601
                      doRunFromMain at org/jruby/Main.java:415
                        internalRun at org/jruby/Main.java:307
                                run at org/jruby/Main.java:234
                               main at org/jruby/Main.java:206

Caused by:
java.lang.ClassNotFoundException: com.apple.eawt.QuitHandler
                          findClass at java/net/URLClassLoader.java:466
                          loadClass at java/lang/ClassLoader.java:566
                          loadClass at java/lang/ClassLoader.java:499
                getDeclaredMethods0 at java/lang/Class.java:-2
          privateGetDeclaredMethods at java/lang/Class.java:3119
                getMethodsRecursive at java/lang/Class.java:3260
                         getMethod0 at java/lang/Class.java:3246
                          getMethod at java/lang/Class.java:2065
                          runSketch at processing/core/PApplet.java:10751
                          runSketch at processing/core/PApplet.java:10989
                          runSketch at processing/core/PApplet.java:10995
                            invoke0 at jdk/internal/reflect/NativeMethodAccessorImpl.java:-2
                             invoke at jdk/internal/reflect/NativeMethodAccessorImpl.java:62
                             invoke at jdk/internal/reflect/DelegatingMethodAccessorImpl.java:43
                             invoke at java/lang/reflect/Method.java:564
  invokeDirectWithExceptionHandling at org/jruby/javasupport/JavaMethod.java:440
                 tryProxyInvocation at org/jruby/javasupport/JavaMethod.java:625
                       invokeDirect at org/jruby/javasupport/JavaMethod.java:301
                               call at org/jruby/java/invokers/InstanceMethodInvoker.java:36
                       cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:323
                               call at org/jruby/runtime/callsite/CachingCallSite.java:139
                        processCall at org/jruby/ir/interpreter/InterpreterEngine.java:344
                          interpret at org/jruby/ir/interpreter/StartupInterpreterEngine.java:72
                          interpret at org/jruby/ir/interpreter/InterpreterEngine.java:78
                   INTERPRET_METHOD at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:136
                               call at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:123
                       cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:334
                               call at org/jruby/runtime/callsite/CachingCallSite.java:149
                        newInstance at org/jruby/RubyClass.java:997
                               call at org/jruby/RubyClass$INVOKER$i$newInstance.gen:-1
                               call at org/jruby/internal/runtime/methods/DynamicMethod.java:194
                               call at org/jruby/java/proxies/ConcreteJavaProxy.java:151
                       cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:323
                               call at org/jruby/runtime/callsite/CachingCallSite.java:139
                        processCall at org/jruby/ir/interpreter/InterpreterEngine.java:344
                          interpret at org/jruby/ir/interpreter/StartupInterpreterEngine.java:72
                          interpret at org/jruby/ir/interpreter/InterpreterEngine.java:78
                   INTERPRET_METHOD at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:136
                               call at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:123
                               call at org/jruby/internal/runtime/methods/DynamicMethod.java:194
                       cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:323
                               call at org/jruby/runtime/callsite/CachingCallSite.java:139
   invokeOther2:load_and_run_sketch at Users/tstuder/$_dot_rvm/gems/ruby_minus_2_dot_5_dot_1/gems/jruby_art_minus_1_dot_5_dot_2/lib/jruby_art/runners//Users/tstuder/.rvm/gems/ruby-2.5.1/gems/jruby_art-1.5.2/lib/jruby_art/runners/run.rb:4
                             <main> at /Users/tstuder/.rvm/gems/ruby-2.5.1/gems/jruby_art-1.5.2/lib/jruby_art/runners/run.rb:4
                invokeWithArguments at java/lang/invoke/MethodHandle.java:707
                               load at org/jruby/ir/Compiler.java:94
                          runScript at org/jruby/Ruby.java:852
                        runNormally at org/jruby/Ruby.java:771
                        runNormally at org/jruby/Ruby.java:789
                        runFromMain at org/jruby/Ruby.java:601
                      doRunFromMain at org/jruby/Main.java:415
                        internalRun at org/jruby/Main.java:307
                                run at org/jruby/Main.java:234
                               main at org/jruby/Main.java:206

```

### JRuby yml
```jruby yml
---
PROCESSING_ROOT: "/Applications/Processing.app/Contents/Java"
JRUBY: false
sketchbook_path: "/Users/tstuder/Documents/Processing"
template: bare
MAX_WATCH: 32
sketch_title: JRubyArt Static Sketch
width: 600
height: 600
```
**Environment**
```
**java --version**
java 10.0.2 2018-07-17
Java(TM) SE Runtime Environment 18.3 (build 10.0.2+13)
Java HotSpot(TM) 64-Bit Server VM 18.3 (build 10.0.2+13, mixed mode)

**ruby --version**
ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-darwin16]

**Processing**
Processing 3.4
```

### Actual Behavior
It appears the file attempts to launch, and then quits.

**Comment by monkstone** [comment]
I can get JRubyArt to run OK on linux with openjdk10, but I have good reason to believe that `com/apple/eawt/QuitHandler` failure has several possible causes:-

1. 

The apple.jar needs to be on the classpath on the Mac (and it isn't).
Java fails to see the proprietary code (which goes back to the last Mac JVM 1.6).
What I can say is that when I attempted to compile java processing with jdk9 it refused to do so because of the proprietary code. You try adding https://github.com/processing/processing/blob/master/core/apple.jar to the classpath possibly by adding it to the lib folder of the JRubyArt gem.

**Comment by tomstuder** [comment]
@monkstone Not a problem. Thank you for reviewing. Please let me know if you need any references to my current setup. I am on mac OS Sierra 10.12.6 (16G1510).

**Comment by monkstone** [comment]
@tomstuder I'm a bit worried by your tweet that you have both processing-2.2.1 and processing-3.4 installed, because I'm not sure how it works on Mac (on most linux installs they can easily co-exist owing to the versioning, except you need two sketchbook folders, typically at /home/tux/sketchbook and /home/sketchbook3). For processing-2.2.1 there is `ruby-processing`, but it's not maintained, `JRubyArt` was created as replacement for `ruby-processing` since processing-3.*. Anyway if the processing installs both work for you it doesn't matter, but what it does mean for you is that you can't rely on my default configurations for the Mac.

`~.rp5rc` for ruby-processing

```yaml
PROCESSING_ROOT: "/Applications/Processing.app/Contents/Java
...
``` 

`.jruby_art/config.yml` for JRubyArt

Note that if you wan't to use an installed jruby instead of jruby-complete set JRUBY: true

```yaml
PROCESSING_ROOT: /Applications/Processing.app/Contents/Java
sketchbook_path: # user defined path, to pick up video/audio libraries etc
JRUBY: false # uses jruby-complete by default for running sketches
template: bare # use class or emacs for alternative templates
```
The PROCESSING_ROOT needs to point to the outer folder where processing has been installed (when we automagically can pick up processing `core.jar` etc and stick them on java classpath) and not symbolic links as it might if you installed using `brew`.

**Comment by monkstone** [comment]
@tomstuder Re jdk10, it is quite possible that jdk8 on MacOS contains the proprietary code (it certainly can help to make awt swing work better on the Mac, since Mac stopped providing there own version of jvm). In which case jdk10 might work with `processing/JRubyArt` if processing.org `apple.jar` was on the classpath.

**Comment by Aerlinger** [comment]
Having the exact same issue with JDK12 on mac OS Sierra 10.14.3. Tried using other versions of the JDK but going in circles and can't seem to get past this issue. Has anyone found a solution?

**Comment by monkstone** [comment]
The current version of processing is not compatible with jdk9+, especially on macOS. Sam Pottinger has a branch that works with jdk11. Further I have compiled and run JRubyArt with it on linux. If you are up to it, it would be great to see if it works on macOS. https://github....sampottinger/processing. Failing that try the latest version of propane.

**Comment by Aerlinger** [comment]
Thanks for the quick response. Using propane appears to work fine for me. 

Is there any disadvantage to using propane vs JRubyArt other than `watch` not being available? I'm not using the processing editor for development and JRubyArt seems to use a bit more "magic" to be similar to processing (i.e. declaring `setup`, `draw`, etc. in the global namespace). I did notice that some of the Propane examples (e.g. [circles](https://github.com/ruby-processing/propane-examples/blob/master/contributed/circles.rb)) are different so wasn't sure which one was best to use.

**Comment by Aerlinger** [comment]
It's also unclear if processing libraries such as Sound/Video can be used from Propane. Looking deeper it looks like the documentation on Propane is almost identical to JRubyArt. 

Is it safe to say that JRubyArt is essentially a wrapper around Propane that abstracts away some of the details of JRuby and also provides an interface to processing?

**Comment by monkstone** [comment]
See how to use the [processing libraries](https://ruby-processing.github.io/propane/libraries.html) and other libraries. For processing libraries they need to [installed](https://ruby-processing.github.io/propane/libraries/processing), this can be done manually as well. JRubyArt is quite different from propane, in that there is two stage load process, this is partly what allows us to support watch mode it is very much a successor to the original ruby-processing, and relies on an installed version of vanilla processing, we pick up processing jar locations (core.jar etc. and libraries jars) via configuration. For propane we compile our own processing core from source and stick it it with our custom code into `propane.jar`. We rely on a path convention to use other processing libraries, so we need to install them ourselves. The `load_library` alias `load_libraries` allows to load pure `ruby` and pure `java` libraries. There`s more certain processing libraries `toxiclibs`, `geomerative` and `wordcram` are available as gems, plus certain [custom libraries](https://ruby-processing.github.io/propane/gems/).

**Comment by monkstone** [comment]
My first response was written quickly on my tablet, so it was briefer that I would like. I would very much like to get to bottom of issue with JRubyArt on macOS. But if we wait for Ben Fry to sort it out, hell may well have frozen over. [Sam Pottinger](https://github.com/sampottinger/processing) has done a brilliant job of updating vanilla processing, and it works on [linux](https://discourse.processing.org/t/arraylist-full-of-functions/10786/6) (but also must work on macOS, because that's what he developed it on. If you are happy with propane all well and good, but make sure to post any issues to https://github.com/ruby-processing/propane/issues.  The current version of JRubyArt should work with Sam Pottingers version of processing (needs ant compile) providing that you point to correct `PROCESSING_HOME`.

**Comment by monkstone** [comment]
OK it seems Ben Fry is going to drag his feet here, because he's no longer paying attention to [Sam Pottingers pull request](https://github.com/processing/processing/pull/5753). A possible solution is compile our own processing core (as was done in early versions of ruby-processing) see https://github.com/monkstone/processing-core.

**Comment by monkstone** [comment]
@tomstuder @Aerlinger is there any chance you could test jruby_art-2.0.0.pre for me, that I have just released to https://rubygems.org/gems/jruby_art, it is compiled with jdk11 and includes fixes that should make it compatible with macOS.
```bash
gem install jruby_art --pre
```

**Comment by monkstone** [comment]
If SamPottingers fix does not work dues this make sense https://alvinalexander.com/source-code/java-macos-about-preferences-quit-handlers-desktop-awt ?

**Comment by sampottinger** [comment]
Hey! Just checking to see what if you had any luck on this issue on your end? Were you able to use my fix?

**Comment by monkstone** [comment]
Hi a @sampottinger I used the code to compile my own processing core. I have yet get confirmation that it works on MacOS. However from @Aerlinger comments propane that uses essentially same code works.

**Comment by monkstone** [comment]
Hey that,s utterly spooky, @benfry with health issues I am currently in hospital awaiting probable pacemaker...

**Comment by sampottinger** [comment]
Ahhh stay healthy processing community! Hope it goes well @monkstone.

**Comment by monkstone** [comment]
@sampottinger thanks for concern, I am now a bionic man having had ICD fitted for ARVCD (I'm the oldest person to present the condition, in the experience of my consultant, further I was completely asymptomatic, so had many tests to diagnose). @Aerlinger @tomstuder could you please try out the latest JRubyArt prelease:-
```bash
gem install jruby_art --pre
```
That should fix issue (NB: move old ~/.jruby_art/config.yml before installing, new config setup)

**Comment by benfry** [comment]
@monkstone, insulting and talking s*t about strangers isn't a great way to motivate them to do things for you. Not sure what I've done to you that deserves this kind of talk:

> I would very much like to get to bottom of issue with JRubyArt on macOS. But if we wait for Ben Fry to sort it out, hell may well have frozen over.

and 

> OK it seems Ben Fry is going to drag his feet here, because he's no longer paying attention to Sam Pottingers pull request. 

Processing has always been a project I do in my spare time, and give away for free. Cut me some slack when I don't have the time, or at least refrain from dragging my name into your rants.

**Comment by monkstone** [comment]
@benfry I hope you are well. I know you are probably as frustrated as I am how Apple and Oracle have made it to develop cross-platform open source projects in java. But I think it has been a long time coming. I am sure there are any number of people that would like to help out, and @sampottinger seems to have made an excellent start.

**Comment by monkstone** [comment]
Should be fixed with JRubyArt-2.0 if it is not please re-open
